internal/input/changes.go

```diff
@@ -116,97 +110,205 @@ func closeChannels(accChannel, offChannel, trustChannel chan xdr.LedgerEntry) {
 	}
 }
 
-// sendLedgerChangesToChannels streams the changes from a single ledger to the provided channels
-func sendLedgerChangesToChannels(seqNum uint32, core *ledgerbackend.CaptiveStellarCore, accChannel, offChannel, trustChannel chan xdr.LedgerEntry) {
-	changeReader, err := ingestio.NewLedgerChangeReader(core, network.PublicNetworkPassphrase, seqNum)
-	if err == nil {
-		compactedChanges := getCompactedChanges(changeReader)
-		for _, change := range compactedChanges {
-			// TODO: figure out how to handle deleted entries
-			if change.Post != nil {
-				entry := *change.Post
-				sendToChannel(entry, accChannel, offChannel, trustChannel)
+func addLedgerChangesToCache(changeReader *ingestio.LedgerChangeReader, accCache, offCache, trustCache *ingestio.LedgerEntryChangeCache) error {
+	for {
+		change, err := changeReader.Read()
+		if err == ingestio.EOF {
+			break
+		}
+
+		if err != nil {
+			return err
+		}
+
+		switch change.Type {
+		case xdr.LedgerEntryTypeAccount:
+			accCache.AddChange(change)
+
+		case xdr.LedgerEntryTypeOffer:
+			offCache.AddChange(change)
+
+		case xdr.LedgerEntryTypeTrustline:
+			trustCache.AddChange(change)
+
+		default:
+		}
+	}
+
+	return nil
+}
+
+// exportBatch gets the changes from the ledgers in the range [batchStart, batchEnd), compacts them, and sends them to the proper channels
+func exportBatch(batchStart, batchEnd uint32, core *ledgerbackend.CaptiveStellarCore, accChannel, offChannel, trustChannel chan ChangeBatch, logger *log.Entry) {
+	accChanges := ingestio.NewLedgerEntryChangeCache()
+	offChanges := ingestio.NewLedgerEntryChangeCache()
+	trustChanges := ingestio.NewLedgerEntryChangeCache()
+	for seq := batchStart; seq < batchEnd; {
+		latestLedger, err := core.GetLatestLedgerSequence()
+		if err != nil {
+			logger.Error("unable to get the lastest ledger sequence: ", err)
+		}
+
+		if seq <= latestLedger {
+			changeReader, err := ingestio.NewLedgerChangeReader(core, password, seq)
+			if err != nil {
+				logger.Error(fmt.Sprintf("unable to create change reader for ledger %d: ", seq), err)
 			}
+
+			err = addLedgerChangesToCache(changeReader, accChanges, offChanges, trustChanges)
+			if err != nil {
+				logger.Error(fmt.Sprintf("unable to read changes from ledger %d: ", seq), err)
+			}
+
+			changeReader.Close()
+
+			// only look at the next ledger if this one was available; otherwise try again until the network closes the ledger
```

Comment:
I get the intention of this comment, but I don't understand why this code would retry. My understanding: it would make a new reader (if the last observed seqnum is before the latest ledger); cache the changes as needed; close the reader; and then increment the seqnum.